mergeRules.py

```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python3
+
+import sys
+import tempfile
+
+from pykWasm import *
+from pykWasm import _notif, _warning, _fatal
+
+################################################################################
+# Direct rule merging                                                          #
+################################################################################
+
+def merge_rules_direct(definition_dir, main_defn_file, main_module, rule_sequences):
+    merged_rules = []
+    for rule_sequence in rule_sequences:
+        gen_rule = mergeRules(definition_dir, main_defn_file, main_module, rule_sequence)
+        if gen_rule is not None:
+            merged_rules.append(gen_rule)
+    return merged_rules
+
+################################################################################
+# Rule Merging based on maximal non-overlapping subsequences                   #
+################################################################################
+
+def maximal_nonoverlapping_subsequences(sequence, subsequence_length = 2):
+    """Calculate maximal length non-overlapping subsequences from a given sequence of strings.
+
+    Input:
+
+        -   sequence: list of strings.
+        -   subsequence_length: size of subsequences to look for.
+
+    Output: maximal non-overlapping subsequences, with earlier subsequences preferred.
+    """
+    subsequences = { }
+
+    for i in range(0, len(sequence) - subsequence_length):
+        subsequence = '|'.join(sequence[ i : i + subsequence_length ])
```

Comment:
Why this?  How do we know this will not appear in an AST?

Reply:
What's being joined here is rule-hashes, not ASTs. Rule-hashes are just `[a-z0-9]*`.